src/clients/blocks.ts

```diff
@@ -1,49 +1,65 @@
 import _ from 'lodash';
 
+import { DBClient, Table } from '../db/db';
+import { Block } from '../types/block';
 import { ChainId } from '../types/chain';
 
 import { EVMClient } from './evm';
 
-const blocksAPI = new GraphQLClient(process.env.BLOCKS_SUBGRAPH_ENDPOINT!);
 
 export type BlocksClient = {
-  fetchBlockTimestamps: (chainId: ChainId, blockNumbers: string[]) => Promise<{ [blockNumber: string]: string }>;
+  fetchBlockTimestamps: (chainId: ChainId, blockNumbers: string[]) => Promise<{ 
+    timestampsByBlockNumber: { [blockNumber: string]: string },
+    newBlocks: Block[] 
+  }>;
 }
 
-const init = async (evmClients: { [chainId in ChainId]: EVMClient }) => {
+const init = async (
+  evmClients: { [chainId in ChainId]: EVMClient },
+  dbClient: DBClient
+) => {
   return {
     fetchBlockTimestamps: async (
       chainId: ChainId,
       blockNumbers: string[],
-    ): Promise<{ [blockNumber: string]: string }> => {
-      let result = {};
-
-      // only using the graph for Ethereum. All other EVM chains will use RPC calls
-      if (chainId === ChainId.ethereum){
-        const blockNumbersForQuery = JSON.stringify(blockNumbers);
-        const { blocks } = await blocksAPI.request(
-          gql`
-          {
-            blocks(first:${blockNumbers.length}, where: {number_in: ${blockNumbersForQuery}}) {
-              number
-              timestamp
-            }
-          }
-          `,
-        );
-        
-        const responseByBlockNumber = _.keyBy(blocks, 'number');
-        const blockTimes = _.mapValues(responseByBlockNumber, response => response.timestamp)
-        result = blockTimes;
-      } else {
-        const blockNumbersParsed: number[] = blockNumbers.map(blockNumber => parseInt(blockNumber))
-
-        const blocks = await evmClients[chainId].getBlockTimestampsByBlockNumber(blockNumbersParsed);
-        result = blocks
+    ): Promise<{ 
+      timestampsByBlockNumber: { [blockNumber: string]: string },
+      newBlocks: Block[] 
+    }> => {
+      let timestampsByBlockNumber: { [blockNumber: string]: string } = {};
+
+      const existingBlocks = await dbClient.getRecords<Block>(Table.blocks, [
+        ['where', ['chainId', chainId]],
+        ['whereIn', ['blockNumber', blockNumbers]]
+      ])
+      console.log(`using ${existingBlocks.length} blocks from cache`)
+      const existingBlocksByBlockNumber = _.keyBy(existingBlocks, 'blockNumber');
+
+      const missingBlockNumbers = blockNumbers.filter(blockNumber => !existingBlocksByBlockNumber[blockNumber]);
+      let newBlocks: Block[] = [];
+
+      const blockNumbersParsed: number[] = missingBlockNumbers.map(blockNumber => parseInt(blockNumber))
+
+      const blocks = await evmClients[chainId].getBlockTimestampsByBlockNumber(blockNumbersParsed);
+
+      newBlocks = Object.keys(blocks).map(blockNumber => ({
+        chainId,
+        blockNumber,
+        timestamp: new Date(parseInt(blocks[blockNumber]) * 1000)
```

Comment:
I didn't know about this. Does it mean that the block number officially corresponds to a distinct point in time, or are we making a bit of an approximation here?

Reply:
nah i think just badly named variables - blocks should be blockTimestamps?